### ISSUE TITLE
Proposal: tuple-friendly return type syntax for methods

### ISSUE THREAD

**Issue by int19h** [dotnet/roslyn#13744]
(taken from NH discussion here: https://news.ycombinator.com/item?id=12361949)

The tuple syntax (#347), while straightforward on its own, is rather distracting in method declarations. E.g. here:
```cs
public (int sum, int count) Tally(IEnumerable<int> values)
```
The name of the method gets "squished" between the return type and the parameter list, even though it's more important than either when scanning code. 

I suggest adding some kind of syntax where the return type follows the parameter list to avoid this problem. There is a precedent for it in a C-style language with traditional declarators, albeit in somewhat different circumstances - C++11 has added the return-type-at-end syntax `auto f(...) -> T`.

Because `->` is already used in C#, and because it is too easily confused with `=>` for lambdas and expression-bodies methods, I don't think C++11 syntax is a good fit. Instead, suggestion is to take the syntax from Ada by reusing `return`: 
```cs
public Tally(IEnumerable<int> values) return (int sum, int count) { ... }
```
or 
```cs
public Tally(IEnumerable<int> values) return (int sum, int count) =>
```
This makes method name the first identifier on the line, inputs next, and finally outputs, matching the natural reading flow ("Tally takes values and returns their sum and count"). It's also more consistent with `Func<TArg, ..., TResult>`.

Grammar-wise, it looks like it's a very straightforward extension on the existing constructor syntax. Basically things that would be parsed as constructors and then rejected because the identifier didn't match the class, would now become methods. Whether to require `return`, or to imply `void` in its absence, is largely a matter of taste. And, obviously, `return` is already a keyword, and it cannot appear in this position today, so this is fully backwards compatible.

I don't suggest applying `return` to anything other than the methods (in particular, not to properties). Tuple types are much less of an issue there from readability perspective because there's no argument list, and thus the name of the property is always in a prominent position at the end of the line, whereas the use of `return` would be confusing for properties with setters:
```cs
public (int sum, int count) Totals {
  get => ...
  set => ...
}
```

While the main motivation is for tuples, there are other circumstances where this is useful. One is any other case where the return type is complicated - for example, a gnarly nested generic:
```cs
Task<Dictionary<string, Func<bool>> Foo(object x) => ...

// becomes

Foo(object x) return Task<Dictionary<string, Func<bool>> => ...
```
Another is when the method itself is generic - as most C# devs know, Intellisense gets in the way when typing out something like:
```cs
public T Foo<T>()
```
because when you get to the first `T`, it's not declared yet, and so Intellisense suggests other identifiers in scope that happen to start with a `T`; you have to explicitly close the completion list to get it out of the way, and if you forget, it'll probably substitute something else when you press space after `T`. On the other hand, with the `return` syntax:
```cs
public Foo<T>() return T
```
Usage follows declaration for `T`, and so you would get proper completion for it in Intellisense by the time you get past `return`.

There's also a marginal benefit in that this syntax permits a more natural way to apply attributes to the return value:
```cs
[return: Foo]
int Bar();

// becomes

int Bar() return [Foo] int;
```
Side note: this presupposes that there's no plan to add a _generic_ form of type-follows-name declarators to C# that would apply everywhere (akin to `x: T` form used by TypeScript and many other languages). If there is such a plan, this is redundant, and `return` is a poor choice, because it couldn't be used uniformly for non-method entities with types.

**Comment by alrz** [comment]
Kinda #8448.

**Comment by int19h** [comment]
Indeed, different syntax for the same purpose. Unfortunately, I didn't find that other item when searching the tracker.

**Comment by alrz** [comment]
By the way, I think `->` operator is a good fit since it is used in C++ for the same purpose(s) and is familiar.
```cs
public F(IEnumerable<int> values) -> (int, int) => (1,2);
```
Feels like Haskell though 😄

**Comment by int19h** [comment]
I personally prefer the `->` actually, but there's a strong argument to be made here that `=>` is already used to designate method/lambda body, and so using `->` to designate the return type will produce confusion, especially among people just learning the language, when both are used in the same line, as in your first snippet. This is confounded by the decision of Java to use `->` for lambdas in lieu of C# `=>`. 

C++ dodges that bullet by not having expression-bodied methods/lambdas (and planning to use a different syntax when they do).

**Comment by gordanr** [comment]
Since this issue was very quickly labeled and assigned, it could be sign that there are already some thoughts among language designers regarding this subject.

I suppose that the main idea is further unification with TypeScript (and F#). Probably mentioned "type-follows-name declarators" augmented with some other related concepts.

Something like:

```
int Calculate(string s, int n, DateTime t) { ... }

function Calculate(s: string, n: int, t: DateTime) : int { ... }
```

I would like to share some very helpful links written by Eric Lippert regarding this topic.

https://blogs.msdn.microsoft.com/ericlippert/2009/05/25/why-is-the-return-type-parameter-last/
http://www.informit.com/articles/article.aspx?p=2425867
http://stackoverflow.com/questions/35966358/why-are-new-programming-languages-shifting-types-to-the-other-side/35980542#35980542

**Comment by Inverness** [comment]
I agree with the sentiment here.

My single issue with the tuple type concept was the mess that it would create in return type declarations and in other cases where the type needs to be specified.

**Comment by laicasaane** [comment]
> Not sure if this should be permitted only for tuple types.

I hope this syntax can be permitted for any method. I had somewhat bad experience when I was working with really long type names before (long names + generics, you can imagine!).

**Comment by paulomorgado** [comment]
When I read C# code, I expect to see types first, not the end.

I'm not opposed to `auto` on private members, but only if the type is inferred.

How come?

Are you looking for Visual Basic? 😄
